File: learnLeetCode/Day39/Part1/game_map_compression.cpp
#include <bits/stdc++.h>
using namespace std;

int min_steps(const vector<string>& grid){
    int n = (int)grid.size();
    if(!n) return -1; int m = (int)grid[0].size(); if(!m) return -1;
    map<pair<int,int>, int> key_id; int kid=0;
    for(int r=0;r<n;++r) for(int c=0;c<m;++c) if(grid[r][c]=='K') key_id[{r,c}] = kid++;
    int all_mask = (1<<kid) - 1;
    auto passable = [&](int r,int c,int mask){
        char ch = grid[r][c];
        if(ch=='#') return false;
        if(ch=='D' && mask!=all_mask) return false;
        return true;
    };
    if(grid[0][0]=='#') return -1;
    int start_mask = 0; if(grid[0][0]=='K') start_mask |= 1<<key_id[{0,0}];
    vector<vector<vector<char>>> vis(n, vector<vector<char>>(m, vector<char>(1<<kid, 0)));
    queue<tuple<int,int,int,int>> q; // r,c,mask,dist
    q.emplace(0,0,start_mask,0); vis[0][0][start_mask]=1;
    int dr[4]={1,-1,0,0}, dc[4]={0,0,1,-1};
    while(!q.empty()){
        auto [r,c,mask,dist]=q.front(); q.pop();
        if(r==n-1 && c==m-1) return dist;
        for(int k=0;k<4;++k){
            int nr=r+dr[k], nc=c+dc[k];
            if(nr<0||nr>=n||nc<0||nc>=m) continue;
            int nmask = mask;
            char ch = grid[nr][nc];
            if(ch=='K') nmask = mask | (1<<key_id[{nr,nc}]);
            if(passable(nr,nc,mask) && !vis[nr][nc][nmask]){
                vis[nr][nc][nmask]=1;
                q.emplace(nr,nc,nmask,dist+1);
            }
        }
    }
    return -1;
}

int main(){ ios::sync_with_stdio(false); cin.tie(nullptr);
    vector<string> grid={".K..","##D.","..#.","...K"};
    cout << min_steps(grid) << "\n";
    return 0; }

File: learnLeetCode/Day39/Part1/game_map_compression.py
from collections import deque
from typing import List

def min_steps(grid: List[str]) -> int:
    n = len(grid)
    m = len(grid[0]) if n else 0
    if n == 0 or m == 0:
        return -1
    # Map keys to bit indices
    key_id = {}
    kid = 0
    for r in range(n):
        for c in range(m):
            if grid[r][c] == 'K':
                key_id[(r,c)] = kid
                kid += 1
    all_mask = (1 << kid) - 1

    def passable(r: int, c: int, mask: int) -> bool:
        ch = grid[r][c]
        if ch == '#':
            return False
        if ch == 'D' and mask != all_mask:
            return False
        return True

    start_mask = 0
    if grid[0][0] == '#':
        return -1
    if grid[0][0] == 'K':
        kid0 = key_id[(0,0)]
        start_mask |= 1 << kid0

    q = deque()
    q.append((0,0,start_mask,0))
    visited = [[[False]*(1<<kid) for _ in range(m)] for _ in range(n)]
    visited[0][0][start_mask] = True

    dirs = [(1,0),(-1,0),(0,1),(0,-1)]
    while q:
        r,c,mask,dist = q.popleft()
        if r == n-1 and c == m-1:
            return dist
        for dr,dc in dirs:
            nr, nc = r+dr, c+dc
            if 0 <= nr < n and 0 <= nc < m:
                nmask = mask
                ch = grid[nr][nc]
                if ch == 'K':
                    kid_idx = key_id[(nr,nc)]
                    nmask = mask | (1<<kid_idx)
                if passable(nr,nc,mask):
                    if not visited[nr][nc][nmask]:
                        visited[nr][nc][nmask] = True
                        q.append((nr,nc,nmask,dist+1))
    return -1

if __name__ == "__main__":
    grid = [
        ".K..",
        "##D.",
        "..#.",
        "...K",
    ]
    print(min_steps(grid))
